crates/sargon/src/signing/petition_types/petitions.rs

```diff
@@ -0,0 +1,262 @@
+        let mut successful_transactions = MaybeSignedTransactions::empty();
+        let mut neglected_factor_sources = IndexSet::<NeglectedFactor>::new();
+        for (intent_hash, petition_of_transaction) in
+            txid_to_petition.into_iter()
+        {
+            let outcome = petition_of_transaction.outcome();
+            let signatures = outcome.signatures;
+
+            if outcome.transaction_valid {
+                successful_transactions.add_signatures(intent_hash, signatures);
+            } else {
+                failed_transactions.add_signatures(intent_hash, signatures);
+            }
+            neglected_factor_sources.extend(outcome.neglected_factors)
+        }
+
+        SignaturesOutcome::new(
+            successful_transactions,
+            failed_transactions,
+            neglected_factor_sources,
+        )
+    }
+
+    pub(crate) fn each_petition<T, U>(
+        &self,
+        factor_source_ids: IndexSet<FactorSourceIDFromHash>,
+        each: impl Fn(&PetitionForTransaction) -> T,
+        combine: impl Fn(Vec<T>) -> U,
+    ) -> U {
+        let for_each = factor_source_ids
+            .clone()
+            .iter()
+            .flat_map(|f| {
+                self.factor_source_to_intent_hashes
+                    .get(f)
+                    .expect("Should be able to lookup intent hash for each factor source, did you call this method with irrelevant factor sources? Or did you recently change the preprocessor logic of the SignaturesCollector, if you did you've missed adding an entry for `factor_source_to_intent_hash`.map")
+                    .iter()
+                    .map(|intent_hash| {
+                        let binding = self.txid_to_petition.borrow();
+                        let value = binding.get(intent_hash).expect("Should have a petition for each transaction, did you recently change the preprocessor logic of the SignaturesCollector, if you did you've missed adding an entry for `txid_to_petition`.map");
+                        each(value)
+                    })
+            }).collect_vec();
+        combine(for_each)
+    }
+
+    pub(crate) fn invalid_transactions_if_neglected_factors(
+        &self,
+        factor_source_ids: IndexSet<FactorSourceIDFromHash>,
+    ) -> IndexSet<InvalidTransactionIfNeglected> {
+        self.each_petition(
+            factor_source_ids.clone(),
+            |p| {
+                p.invalid_transaction_if_neglected_factors(
+                    factor_source_ids.clone(),
+                )
+            },
+            |i| i.into_iter().flatten().collect(),
+        )
+    }
+
+    pub(crate) fn should_neglect_factors_due_to_irrelevant(
+        &self,
+        factor_sources_of_kind: &FactorSourcesOfKind,
+    ) -> bool {
+        let ids = factor_sources_of_kind
+            .factor_sources()
+            .iter()
+            .map(|f| f.factor_source_id().as_hash().unwrap().clone()) // TODO change that
```

Comment:
TODO:
The same should apply here.

Reply:
you can change `unwrap()` to `expect("Signature Collector only works with HD FactorSources. And ALL HD FactorSource should have FactorSourceIDFromHash type as ID")`

Reply:
Right thanks @CyonAlexRDX